Msi 2382 graph ql support for in store pickup mutation

swnsma opened the pull request:
## Description
Extension of Quote Graph Ql endpoint to become able to pass pickup location code.
Issue: #2382

Commit:
MSI-2382 GraphQl support for In-Store Pickup [Mutation].
Add simple module declaration.

Commit:
MSI-2382 GraphQl support for In-Store Pickup [Mutation].
Add parameter to SetShippingAddress mutation.

Commit:
MSI-2382 GraphQl support for In-Store Pickup [Mutation].
Add api tests coverage.

Commit:
MSI-2382 GraphQl support for In-Store Pickup [Mutation].
Fix static tests.

swnsma commented:
Not sure about copy-paste detector.
For me, it looks that it will be easier to maintain, if the code will be the same.

ishakhsuvarov commented:
@swnsma CPD is something we can ignore if there is a valid reason. Shouldn't be a problem here.

Commit:
MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation.
Fix bug with deletion of address pickup locatio .

Commit:
MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation.
Add Pickup Location to Quote Shipping Address Graph Ql Schema.
Fix adding of Pickup Location Code to Shipping Address.

Commit:
Merge branch 'store-pickup' into MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation

Commit:
Fix php version in composer.json

swnsma committed:
Update schema.graphqls

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Test/_files/create_in_store_pickup_quote_on_eu_website.php:
```diff
@@ -112,3 +113,9 @@
 $cartRepository->save($cart);
+
+/** @var \Magento\Quote\Model\QuoteIdMask $quoteIdMask */
+$quoteIdMask = Bootstrap::getObjectManager()->create(QuoteIdMaskFactory::class)->create();
```
Should we specify in the fixture name that this is a guest quote?

swnsma replied:
To be honest, I think there need to be done arrangement of fixtures.
I think we can rename it to 'guest' but during API tests coverage, fixtures need to be refactored.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Test/_files/create_in_store_pickup_quote_on_eu_website_customer.php:
```diff
@@ -0,0 +1,119 @@
+$cartRepository->save($cart);
+
+/** @var \Magento\Quote\Model\QuoteIdMask $quoteIdMask */
+$quoteIdMask = Bootstrap::getObjectManager()->create(QuoteIdMaskFactory::class)->create();
```
Fixture name states that it's intended for customer, but this seems to be another guest quote.

swnsma replied:
The point that `masked_quote_id` is required in any case.
Please see.
`\Magento\QuoteGraphQl\Model\Resolver\SetShippingAddressesOnCart::resolve`
`\Magento\QuoteGraphQl\Model\Cart\GetCartForUser::execute`

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuote/Model/Quote/ValidationRule/InStorePickupQuoteValidationRule.php:
```diff
@@ -85,10 +85,15 @@ public function validate(Quote $quote): array
-        if (!$this->isPickupLocationShippingAddress->execute($pickupLocation, $address)) {
+        if (!$pickupLocation) {
+            $validationErrors[] = __(
+                'Pickup Location is missed for In-Store Pickup Quote.'
```
```suggestion
                'Quote does not have Pickup Location assigned.'
```

swnsma replied:
Agree.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuote/Model/ResourceModel/DeleteQuoteAddressPickupLocation.php:
```diff
@@ -41,6 +41,6 @@ public function execute(int $addressId): void
         $table = $this->connection->getTableName('inventory_pickup_location_quote_address');
 
-        $connection->delete($table, [self::ADDRESS_ID => $addressId]);
+        $connection->delete($table, [self::ADDRESS_ID . ' in (?)' => $addressId]);
```
Should this method be modified to accept multiple address id's?

swnsma replied:
I think no.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuote/Plugin/Quote/PassPickupLocationToOrder.php:
```diff
@@ -7,14 +7,12 @@
- * @TODO Move logic to fieldset.xml when issue will be resolved in core..
- * @see Please check issue in core for more details: https://github.com/magento/magento2/issues/23386.
  */
 class PassPickupLocationToOrder
```
```suggestion
 class SetPickupLocationCodeToOrder
```

swnsma replied:
Agree.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuoteGraphQl/Model/Cart/SetShippingAddressesOnCart.php:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Set shipping address to the cart. Proceed with passed Pickup Location code.
+ */
+class SetShippingAddressesOnCart implements SetShippingAddressesOnCartInterface
```
```suggestion
class SetShippingAddressesToCart implements SetShippingAddressesToCartInterface
```

swnsma replied:
The original interface has name 'SetShippingAddressesOnCartInterface'.
I think that this is better to keep them similar.

ishakhsuvarov committed:
Merge branch 'store-pickup' of github.com:magento/inventory into MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation

Conflicts:
	composer.lock

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupQuoteGraphQl/Model/Cart/SetShippingAddressesOnCart.php:
```diff
@@ -0,0 +1,143 @@
+     * @throws GraphQlInputException
+     * @throws GraphQlNoSuchEntityException
+     */
+    private function getShippingAddress(array $shippingAddressInput, ContextInterface $context): Address
```
Why isn't it possible just to implement a plugin for the class where this logic originally came from?

swnsma replied:
Do you mean `around` plugin without calling the `proceed`?

lenaorobei commented on app/code/Magento/InventoryInStorePickupQuoteGraphQl/Test/Api/Customer/PickupLocationForShippingTest.php:
```diff
@@ -0,0 +1,316 @@
+    }
+  ) {
+    cart {
+      shipping_addresses {
```
Is it expected behaviour that pickup location address is considered as shipping address? Maybe I'm not in the context.

ishakhsuvarov replied:
Yes, quote must have Pickup Location shipping address instead of customer's address.

lenaorobei commented on app/code/Magento/InventoryInStorePickupQuoteGraphQl/Test/Api/Guest/PickupLocationForShippingTest.php:
```diff
@@ -0,0 +1,172 @@
+  setShippingAddressesOnCart(
+    input: {
+      cart_id: "$mask"
+      shipping_addresses: [
```
What is the purpose of sending shipping address here?

swnsma replied:
1. Check that address will be replaced with Pickup Location address.
2. Some of attributes will not be replaced with Pickup Location address (like telephone).
But these attributes are required for Quote Address.

lenaorobei commented on app/code/Magento/InventoryInStorePickupQuoteGraphQl/etc/schema.graphqls:
```diff
@@ -0,0 +1,10 @@
+}
+
+type ShippingCartAddress implements CartAddressInterface {
+    pickup_location_code: String @resolver(class: "\\Magento\\InventoryInStorePickupQuoteGraphQl\\Model\\Resolver\\ShippingAddress\\PickupLocationCode")
```
Probably we need `cart_items` here if expected behaviour is similar to `type ShippingCartAddress implements CartAddressInterface`, see 
https://github.com/magento/inventory/blob/72f3e99016ce2020e972482bcdd85634933fd252/app/code/Magento/QuoteGraphQl/etc/schema.graphqls#L219

swnsma replied:
@lenaorobei,
I am not sure if I correctly understand your request.
I have just tried to receive 'cart_items' with GraphQl and it looks like working without extra changes.
As I understand, the field from `app/code/Magento/InventoryInStorePickupQuoteGraphQl/etc/schema.graphqls` has been merged to fields from `app/code/Magento/QuoteGraphQl/etc/schema.graphqls`.

Or we are talking about different things?

ishakhsuvarov reviewed (changes requested):
Please take a look at the review comments

Commit:
Msi 2382 graph ql support for in store pickup mutation.
Apply code review changes.

Commit:
Merge branch 'MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation' of github.com:magento-engcom/msi into MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation

swnsma commented:
Hi guys, @ishakhsuvarov @lenaorobei,

Thank you for your review.
I have reacted on code review notices.

ishakhsuvarov committed:
Merge branch 'store-pickup' of github.com:magento/inventory into MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation

Commit:
Merge branch 'store-pickup' into MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation

Commit:
MSI-2382 GraphQl support for In-Store Pickup [Mutation].
Fix Pickup Location missed for created order.

Commit:
Merge branch 'MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation' of github.com:magento-engcom/msi into MSI-2382-GraphQl-support-for-In-Store-Pickup-Mutation

lenaorobei commented:
No other comments from my side. Please make sure that schema is submitted and approved here https://github.com/magento/architecture/tree/master/design-documents/graph-ql/coverage

ishakhsuvarov reviewed (approved):
I've filed an issue https://github.com/magento/inventory/issues/2557 to take one more look at the copy-pasted code in the future
